plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -387,8 +406,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
       // path to write the snapshots to
       bfs::path _snapshots_dir;
 
-      // max time in milliseconds that a read-only transaction is allowed to run
-      int32_t _max_read_only_transaction_time_ms = 150;
+      // ro for read-only
+      struct ro_trx_t {
+         packed_transaction_ptr  trx;
+         next_func_t             next;
+      };
+      // The queue storing read-only transactions to be executed by read-only threads
+      struct ro_trx_queue_t {
+         std::mutex              mtx;
+         std::deque<ro_trx_t>    queue;
+      };
+      uint16_t                        _ro_thread_pool_size{ 0 };
+      static constexpr uint16_t       _ro_max_eos_vm_oc_threads_allowed{ 8 }; // Due to uncertainty to get total virtual memory size on a 5-level paging system, set a hard limit
+      named_thread_pool<struct read>  _ro_thread_pool;
+      fc::microseconds                _ro_write_window_time_us{ 200000 };
+      fc::microseconds                _ro_read_window_time_us{ 60000 };
+      static constexpr fc::microseconds _ro_read_window_minimum_time_us{ 10000 };
+      fc::microseconds                _ro_read_window_effective_time_us{ 0 }; // calculated during option initialization
+      boost::asio::deadline_timer     _ro_write_window_timer;
+      boost::asio::deadline_timer     _ro_read_window_timer;
+      fc::microseconds                _ro_max_trx_time_us{ 0 }; // calculated during option initialization
+      ro_trx_queue_t                  _ro_trx_queue;
+      std::atomic<int32_t>            _ro_num_active_trx_exec_tasks{ 0 };
```

Comment:
Why did you make this change to `int32_t`. It should not be possible to go negative and it makes the `EOS_ASSERT` less strict, not more strict.

Reply:
I was thinking to protect against more than one read-only threads doing the `--_ro_num_active_trx_exec_tasks`. But now I realize _ro_num_active_trx_exec_tasks starting with number of the threads. It would never go down below 0. Will revert to uint32_t but keep checking both the number to be 0 and the task queue to be empty.

Reply:
It protects against that now because you do `== 0` and a `--` on `0` would give you max `uint32_t` which would fail the assert.